nemo/collections/speechlm/models/speech_to_text_llm_model.py

```diff
@@ -0,0 +1,1127 @@
+            input_len_list = [start_idx_list_i[j + 1] - start_idx_list_i[j] for j in range(len(start_idx_list_i) - 1)]
+            input_emb_list = input_embeds[i].split(input_len_list)
+            encoder_input_i = [input_emb_list[0]]
+            for j in range(1, len(input_emb_list)):
+                encoder_input_i.append(encoded[i][j - 1][: encoded_len[i][j - 1]])
+                encoder_input_i.append(input_emb_list[j])
+            encoder_input_i = torch.cat(encoder_input_i)  # T, C
+            encoder_length_i = encoded_len[i].sum() + input_length[i]  # total length of audio and text features
+            max_length = max(max_length, encoder_input_i.size(0))
+            encoder_input_list.append(encoder_input_i)
+            encoder_length_list.append(encoder_length_i)
+
+        encoder_input = torch.stack(
+            [torch.nn.functional.pad(f, (0, 0, 0, max_length - f.size(0))) for f in encoder_input_list]
+        )
+        encoder_length = torch.LongTensor(encoder_length_list).to(encoder_input.device)
+        return encoder_input, encoder_length
+
+    def inject_perception_input(
+        self,
+        encoded: Union[torch.Tensor, List[torch.Tensor]],
+        encoded_len: Union[torch.Tensor, List[torch.Tensor]],
+        input_ids: torch.Tensor,
+        input_length: torch.Tensor,
+        context_start_idx: Optional[List[List[int]]] = None,
+    ):
+        """Inject audio features into the text input and return the final input embeddings to LLM."""
+        # [b, t, c]
+        lm_embedding = self.language_model.embedding
+        input_embeds = lm_embedding.word_embeddings(input_ids)
+
+        if isinstance(encoded, torch.Tensor):
+            # single audio
+            combined_embed, combined_embed_length = self._concat_features(
+                encoded, encoded_len, input_embeds, input_length
+            )
+        else:
+            # concat multiple audios with text segments
+            combined_embed, combined_embed_length = self._concat_multi_features(
+                encoded, encoded_len, input_embeds, input_length, context_start_idx
+            )
+
+        attention_mask = self._create_attention_mask(combined_embed)
+        position_ids = build_position_ids(combined_embed[:, :, 0])
+
+        # Add position embeddings
+        if (
+            getattr(lm_embedding, "position_embeddings", None) is not None
+            and lm_embedding.position_embedding_type == 'learned_absolute'
+        ):
+            position_embeddings = lm_embedding.position_embeddings(position_ids)
+            combined_embed = combined_embed + position_embeddings
+
+        encoder_max_length = combined_embed.shape[1]
+        # Data format change to avoid explicit tranposes : [b s h] --> [s b h].
+        combined_embed = combined_embed.transpose(0, 1).contiguous()
+
+        # If the input flag for fp32 residual connection is set, convert for float.
+        if lm_embedding.config.fp32_residual_connection:
```

Comment:
... how did you even find out that you had to do this step i wonder

Reply:
by looking into the implementation in nlp and mcore....I did spend some time to figure it out